### ISSUE TITLE
[sort-type-constituents] Option for array order

### ISSUE THREAD

**Issue by bordecal** [typescript-eslint/typescript-eslint#2937]
When a type and an array of the type are included in a union (e.g.: `string | string[]`), the `sort-type-union-intersection-members` rule currently flags a violation when the array comes second.  It would be quite convenient to be able to specify a preferred sorting order for arrays, particularly given that most code readers probably find `string | string[]` to be more readable than `string[] | string`.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/sort-type-union-intersection-members": "error"
  }
}
```

```TS
addHttpHeaders(target: HttpHeaders, additionalHeaders: Record<string, string | string[]>) {
    // ...
}
```

**Expected Result**

The `sort-type-union-intersection-members` would not emit a violation for `string | string[]` (or would at least be configurable not to emit a violation).

**Actual Result**

A violation is emitted, and it does not appear to be possible to configure the rule to avoid this.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.13.0` |
| `@typescript-eslint/parser`        | `4.13.0` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.17.0` |
| `node`                             | `14.15.4` |

**Comment by reduckted** [comment]
I also find the default order a bit strange. I expected it to be alphabetical like it says in the docs.

> Enforces that members of a type union/intersection are sorted alphabetically

But `string` comes _before_ `A`, and `readonly string[]` comes _after_ `string[]`. 😕 

Personally, I would like everything to be sorted alphabetically, _except_ for `null` and `undefined`, which should be last.

I wonder if that could be achieved by combining multiple groups. Something like this, perhaps?

```js
groupOrder: [
    [
        'named',
        'keyword',
        'operator',
        'literal',
        'function',
        'import',
        'conditional',
        'object',
        'tuple',
        'intersection',
        'union'
    ],
    'nullish'
]
```
  ],

**Comment by karlhorky** [comment]
I also mentioned this just now over in my question to the ESLint Stylistic maintainers about potentially migrating this rule to ESLint Stylistic:

- https://github.com/typescript-eslint/typescript-eslint/issues/8792#issuecomment-2025344673
